### PR TITLE
Implemented `MainMenu::DrawStartMenu`

### PR DESCRIPTION
**Pull request by PieVieRo**
the only thing that's different is the jump table offset (which obviously will be different for now)

### CODE REVIEW

File: src/MainMenu.cpp [hunk 1]
```diff
@@ -113,3 +117,172 @@ ZunResult MainMenu::LoadTitleAnm(MainMenu *menu)
...
+        this->cursor += i;
+    }
+    AnmVm *drawVm = this->vm;
+    for (i = 0; i < 8; i++, drawVm++ /* zun why */)
+    {
+        DrawMenuItem(drawVm, i, this->cursor, 0xffff0000, 0x80300000, 122);
+    }
+    if (this->stateTimer >= 0x14)
+    {
+        if (WAS_PRESSED(0x1001))
```
Comment (roblabla) [inline src/MainMenu.cpp:139]: It'd be nice to have some defines for the key. This is likely `KEY_ENTER | KEY_SHOOT`.

We have a comment in GetInput:

```
    GetKeyboardState((PBYTE)keyboardState);
                    /* 38  = VK_UP = KEY_UP
                       40  = VK_DOWN = KEY_DOWN
                       37  = VK_LEFT = KEY_LEFT
                       39  = VK_RIGHT = KEY_RIGHT
                       104 = VK_NUMPAD8 = KEY_UP
                       98  = VK_NUMPAD2 = KEY_DOWN
                       100 = VK_NUMPAD4 = KEY_LEFT
                       102 = VK_NUMPAD6 = KEY_RIGHT
                       103 = VK_NUMPAD7 = KEY_LEFT | KEY_UP
                       105 = VK_NUMPAD9 = KEY_RIGHT | KEY_UP
                       97  = VK_NUMPAD1 = KEY_LEFT | KEY_DOWN
                       99  = VK_NUMPAD3 = KEY_RIGHT | KEY_DOWN
                       36  = VK_HOME = KEY_HOME
                       90  = Z = KEY_BOMB
                       88  = X = KEY_BOMB
                       16  = VK_SHIFT = KEY_FOCUS
                       27  = VK_ESCAPE = KEY_MENU
                       17  = VK_CONTROL = KEY_SKIP
                       81  = Q = KEY_Q
                       83  = S = KEY_S
                       13  = VK_RETURN = KEY_ENTER */
                    /* This sets a bitfield of which buttons are pressed.
                       
                       0x001 is shoot
                       0x002 is BOMB
                       0x004 is FOCUS
                       0x008 is MENU
                       0x010 is UP
                       0x020 is DOWN
                       0x040 is LEFT
                       0x080 is RIGHT
                       0x100 is SKIP
                       0x200 is KEY_Q
                       0x400 is KEY_S
                       0x1000 is KEY_ENTER */
```

File: src/MainMenu.cpp [hunk 2]
```diff
@@ -113,3 +117,172 @@ ZunResult MainMenu::LoadTitleAnm(MainMenu *menu)
...
+        {
+            switch (this->cursor)
+            {
+            case 0:
+                for (i = 0; i < 122; i++)
```
Comment (roblabla) [inline src/MainMenu.cpp:144]: Instead of hardcoding 122 here, use `ARRAY_SIZE(this->vm)`

File: src/MainMenu.cpp [hunk 3]
```diff
@@ -113,3 +117,172 @@ ZunResult MainMenu::LoadTitleAnm(MainMenu *menu)
...
+                this->stateTimer = 0;
+                for (i = 0; i < 122; i++)
+                {
+                    this->vm[i].pendingInterrupt = 4;
+                }
+                g_SoundPlayer.PlaySoundByIdx(0xb, 0);
+                break;
+            }
+        }
+        if (WAS_PRESSED(0x200))
```
Comment (roblabla) [inline src/MainMenu.cpp:270]:
```suggestion
        if (WAS_PRESSED(KEY_Q))
```

File: src/MainMenu.cpp [hunk 4]
```diff
@@ -113,3 +117,172 @@ ZunResult MainMenu::LoadTitleAnm(MainMenu *menu)
...
+        {
+            this->gameState = STATE_QUIT;
+            this->stateTimer = 0;
+            for (i = 0; i < 122; i++)
+            {
+                this->vm[i].pendingInterrupt = 4;
+            }
+            g_SoundPlayer.PlaySoundByIdx(0xb, 0);
+        }
+        if (WAS_PRESSED(0xA))
```
Comment (roblabla) [inline src/MainMenu.cpp:280]:
```suggestion
        if (WAS_PRESSED(KEY_BOMB | KEY_MENU))
```

File: src/MainMenu.cpp [hunk 5]
```diff
@@ -113,3 +117,172 @@ ZunResult MainMenu::LoadTitleAnm(MainMenu *menu)
...
+                }
+                if (EXTRA <= g_Supervisor.cfg.defaultDifficulty)
+                {
+                    g_Supervisor.cfg.defaultDifficulty = NORMAL;
+                }
+                this->stateTimer = 0;
+                this->unk_81fc = 0x40000000;
+                this->maybeMenuTextColor = 0xff000000;
+                this->unk_820c = 0;
+                this->isActive = 60;
```
Comment (roblabla) [inline src/MainMenu.cpp:162]: Hmm, this name is a bit unfortunate. I usually expect a variable named `isX` to be a boolean, containing only 0/1. Do we know what this is used for, and could we find a better name for it? If not, that's fine, we can always delegate this to the future.

**Review by roblabla — CHANGES_REQUESTED** [review]
Couple nitpicks, but this is really solid!

File: src/MainMenu.cpp (hunk 1 above)
Comment (PieVieRo) [reply src/MainMenu.cpp:139]: Ohhh, I didn't know about those, looks like I haven't familiarised myself with the codebase as much as I thought

File: src/MainMenu.cpp (hunk 5 above)
Comment (PieVieRo) [reply src/MainMenu.cpp:162]: That's the funny thing, even thought it's not a bool, it's only ever set to either 60 or 0
the comment from the ghidra server (not by me) says

> Seems to only get set once the current menu is active (aka the function has been called), 0 when its not ready (loading, or on the title screen), and 60 when it is ready and active.

So like, the tldr is that it's not a bool de iure but it is de facto

File: src/MainMenu.cpp [hunk 6]
```diff
@@ -113,3 +117,172 @@ ZunResult MainMenu::LoadTitleAnm(MainMenu *menu)
...
+    i = MoveCursor(this, 8);
+    if ((this->cursor == 1) && !g_GameManager.hasReachedMaxClears(0, 0) && !g_GameManager.hasReachedMaxClears(0, 1) &&
+        !g_GameManager.hasReachedMaxClears(1, 0) && !g_GameManager.hasReachedMaxClears(1, 1))
+    {
+        this->cursor += i;
+    }
+    AnmVm *drawVm = this->vm;
+    for (i = 0; i < 8; i++, drawVm++ /* zun why */)
+    {
+        DrawMenuItem(drawVm, i, this->cursor, 0xffff0000, 0x80300000, 122);
```
[the remark this answers is not in the document]
Comment (PieVieRo) [reply src/MainMenu.cpp:135]: honestly I don't know, the parameter's name on the ghidra is `vm_amount` but from the usage it looks like an offset in the sprite index table (the only line that references it is `AnmManager::SetActiveSprite(g_AnmManager,vm,vm->anotherSpriteNumber + vm_amount);`, where the last parameter is a sprite index)

File: src/MainMenu.cpp (hunk 5 above)
Comment (roblabla) [reply src/MainMenu.cpp:187]: Alright, let's keep it that way then

File: src/MainMenu.cpp (hunk 1 above)
Comment (roblabla) [reply src/MainMenu.cpp]: To be fair, I literally just finished figuring those out :P

File: src/MainMenu.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (roblabla) [reply src/MainMenu.cpp]: Alright let's keep it hardcoded, we can figure it out later.

**Pull request merged**